### PR TITLE
Add Jira notifier.

### PR DESCRIPTION
**Pull request by devnev**

### CODE REVIEW

File: config/notifiers.go [hunk 1]
```diff
@@ -392,3 +403,54 @@ func (c *PushoverConfig) UnmarshalYAML(unmarshal func(interface{}) error) error
...
+
+	Project     string                 `yaml:"project" json:"project"`
+	IssueType   string                 `yaml:"issueType" json:"issueType"`
+	Priority    string                 `yaml:"priority" json:"priority"`
+	Summary     string                 `yaml:"summary" json:"summary"`
+	Description string                 `yaml:"description" json:"description"`
+	Fields      map[string]interface{} `yaml:"fields" json:"fields"`
+
+	// Catches all undefined fields and must be empty after parsing.
+	XXX map[string]interface{} `yaml:",inline" json:"-"`
```
Comment (bwplotka) [inline config/notifiers.go:422]: Should be comma before inline?

File: config/notifiers.go [hunk 2]
```diff
@@ -392,3 +403,54 @@ func (c *PushoverConfig) UnmarshalYAML(unmarshal func(interface{}) error) error
...
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *JiraConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultJiraConfig
+	type plain JiraConfig
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+	if c.VSendResolved {
+		return fmt.Errorf("sending resolved notifications to Jira is not supported")
```
Comment (bwplotka) [inline config/notifiers.go:433]: Why? Cannot we close the tickets?

File: notify/impl.go [hunk 3]
```diff
@@ -934,6 +938,79 @@ func (n *Pushover) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
...
+	data := n.tmpl.Data(receiverName(ctx), groupLabels(ctx), as...)
+	tmpl := tmplText(n.tmpl, data, &err)
+
+	fields := map[string]interface{}{
+		"project":     map[string]string{"key": tmpl(n.conf.Project)},
+		"issuetype":   map[string]string{"name": tmpl(n.conf.IssueType)},
+		"summary":     tmpl(n.conf.Summary),
+		"description": tmpl(n.conf.Description),
+	}
+	// check errors from tmpl()
```
Comment (bwplotka) [inline notify/impl.go:973]: Why not immediately after this weird tmplText function? Then you can drop the comment

File: template/default.tmpl [hunk 4]
```diff
@@ -190,3 +190,6 @@ Alerts Resolved:
 {{ end }}
 {{ end }}
 {{ define "pushover.default.url" }}{{ template "__alertmanagerURL" . }}{{ end }}
+
+{{ define "jira.default.summary" }}{{ template "__subject" . }}{{ end }}
```
Comment (bwplotka) [inline template/default.tmpl:194]: Not sure how this ticket title will look like... How we could distinguish that from the manually created ticket?

**Review by bwplotka — COMMENTED** [review]
Looks good, some question though

File: config/notifiers.go (hunk 1 above)
Comment (devnev) [reply config/notifiers.go:422]: no, the `inline` is an option, not the field name. see other such structs in this file.

File: config/notifiers.go (hunk 2 above)
Comment (devnev) [reply config/notifiers.go:433]: a) there's no way of keeping state like the ticket ID associated with an alert. the relationship between firing and resolved alert is via the groupkey. I would have to associate the ticket with the group key somehow (e.g. a label), which is more complex than I'd like to add right now
b) I don't really want tickets to be auto-closed. if it fired, it requires investigation.

File: notify/impl.go (hunk 3 above)
Comment (devnev) [reply notify/impl.go:973]: `tmplText()` returns a function which I'm storing in `tmpl`. the `tmpl` function, when called, updates the err object passed into `tmplText`. So i'm checking the result of the `tmpl` calls here, not `tmplText`.

File: template/default.tmpl (hunk 4 above)
Comment (devnev) [reply template/default.tmpl:194]: i'll show you the test ticket. the default __subject is quite clear that its an alert, although we'll probably configure ours to be a little more readable.

File: notify/impl.go (hunk 3 above)
Comment (bwplotka) [reply notify/impl.go:973]: ah sorry, did not see that, cool

**Review by bwplotka — APPROVED** [review]
Let's think how we can avoid DUPLICATES

**Comment by devnev** [comment]
> Let's think how we can avoid duplicates

I'm leaving that to future improvements as it requires a bit more thought about ticket states and some way of associating tickets with alerts.

**Pull request merged**